docs/internals/layout_in_memory.rst

```diff
@@ -5,36 +5,121 @@
 Layout in Memory
 ****************
 
+Reserved Memory Areas
+=====================
+
 Solidity reserves four 32-byte slots, with specific byte ranges (inclusive of endpoints) being used as follows:
 
 - ``0x00`` - ``0x3f`` (64 bytes): scratch space for hashing methods
-- ``0x40`` - ``0x5f`` (32 bytes): currently allocated memory size (aka. free memory pointer)
+- ``0x40`` - ``0x5f`` (32 bytes): allocated memory size (aka. free memory pointer)
 - ``0x60`` - ``0x7f`` (32 bytes): zero slot
 
-Scratch space can be used between statements (i.e. within inline assembly). The zero slot
-is used as initial value for dynamic memory arrays and should never be written to
-(the free memory pointer points to ``0x80`` initially).
+Memory Allocation
+-----------------
+
+Solidity compiles to EVM instructions, and implements its complex data types on top of
+EVM features. Memory allocation is not an EVM feature. Inestead, Solidity implements its
+memory allocation via the free memory pointer stored at location ``0x40``. That value tracks the next
+unallocated memory address available.
+The Solidity compiler generates code that always places new objects at the location pointed to by
+the free memory pointer.
+In inline assembly memory can be allocated manually by increasing the value of the pointer.
+
+Memory located beyond the value of the free memory pointer is considered scratch memory.
+Access to memory in inline assembly is not restricted by the value of the free memory pointer
+and can freely refer to memory located beyond it.
+
+Allocated memory in Solidity is never freed.
+In other words, the value of the free memory pointer is never reduced by Solidity.
+This might change in future versions.
+
+Active Memory and Memory Expansion
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The concept of *allocated memory* is distinct and separate from *active memory*.
```

Comment:
Is the term "active memory" used anywhere else?

Reply:
Yep. It's actually in the Yellow Paper :)